### DISCUSSION TITLE
ESLint in Next.js and Create Next App

### DISCUSSION THREAD

**Discussion by housseindjirdeh** [vercel/next.js#24900] [Ideas]
**Over 50%** of Next.js apps use [ESLint](https://eslint.org/) in some part of their build workflow. This discussion covers steps taken to include a Next.js' specific plugin and shareable config and proposes a default setup for all new applications built with Create Next App.

### tl.dr:

- [`eslint-plugin-next`](https://www.npmjs.com/package/@next/eslint-plugin-next) is an ESLint plugin that aims to catch common issues and problems in a Next.js application. There are a list of initial rules that this plugin will provide, but we would love to hear more suggestions.
- [`eslint-config-next`](https://www.npmjs.com/package/eslint-config-next/v/10.2.1-canary.3) is a shareable config that bundles multiple useful plugins (`next`, `react`, and `react-hooks`) and can easily be included in any Next.js application’s `.eslintrc` file with `({ "extends": "next"})`
- **Proposals:**
  - Include a stricter, separate entry-point for the config (`next/core-web-vitals`) to error on a few, important rules that will affect [Core Web Vitals](https://web.dev/vitals/)
  - ESLint will be enabled at build time by default in Create Next App and both `next` and `next/core-web-vitals` will be included into the initial config

---------------

# Plugin

The following is an initial set of rules that [`eslint-plugin-next`](https://www.npmjs.com/package/@next/eslint-plugin-next) will check for:

#### Imports

- `next/document` should not be imported outside of `pages/_document.js` (#24832)
- `next/head` should not be imported in `pages/_document.js` (#24832)

#### Images

- Do not use `<img>`. Use `<Image>` instead. (#23402)

#### Google Fonts

- `font-display` value is missing (or `fallback`/`block` is used) for Google Fonts (#24766)
- `rel="preconnect"` is missing for Google Fonts domain (#24766)

#### Polyfills

- No duplicate polyfills from Polyfill.io are allowed (#15277)

#### Links

- Do not use the HTML `<a>` tag to navigate to a page. Use `Link` from `next/link` instead (#12969)
- `passHref=true` must be assigned when using `Link` without an HTML `<a>` element for a custom component (#24670)

#### Scripts

- Synchronous scripts are forbidden (#11986)

#### Styles

- Do not include stylesheets manually (#11986)

#### Head

- `<title>` should not be used in `<Head>` in  _document.js (#24868)

# Shareable Config

New, experimental ESLint changes have been introduced to Next.js ([still experimental/in canary](https://github.com/vercel/next.js/pull/23702)). When `{ experimental: { eslint: true } }` is now enabled in `next.config.js`:

1. ESLint will run on every build (`next build`).
2. Creating a new `.eslintrc` file for the first time and running a build will automatically populate the file to extend from [`eslint-config-next`](https://www.npmjs.com/package/eslint-config-next/v/10.2.1-canary.3) (the new shareable config):

```
{
    "extends": "next"
}
```

This config extends recommended rule sets from [`eslint-plugin-react`](https://github.com/yannickcr/eslint-plugin-react), [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks), and [`eslint-plugin-next`](https://www.npmjs.com/package/@next/eslint-plugin-next).

# Proposals

Although this discussion aims to start a conversation on how ESLint is currently configured within Next.js, it also proposes the following:

* Create an additional ESLint config entry point (`next/core-web-vitals`) that can be used by adding the following to `.eslintrc`:

  ```
  {
      "extends": ["next", "next/core-web-vitals"]
  }
  ```

  This would be stricter than the recommended rule-set (which is almost all warnings) and error on a few important rules:
  
  * **_Error:_** Do not use `<img>`. Use `<Image>` instead.
  * **_Error:_** Synchronous scripts are forbidden
  * **_Error:_** Do not use the HTML `<a>` tag to navigate to a page. Use `Link` from `next/link` instead.
  
  If any of these rules error with ESLint enabled, the **build will fail**. As more rules get added to the plugin, only those that are guaranteed to negatively affect Core Web Vital metrics will be included into this entry point.

* By default, ESLint will enabled in Create Next App with the following default `.eslintrc` configuration:

  ```
  {
      "extends": ["next", "next/core-web-vitals"]
  }
  ```

  This will be set up to be easily opted-out of if need be and the default config can always be modified by the developer.

Thoughts? Suggestions? Please share!

**Comment by Xetera** [comment]
I don't believe the `<Image>` check should be turned on by default. The state of next.js image optimization seems to be in a pretty contentious #19914 spot right now and there are many other reasons for why you might not want image optimization, especially if you serve images through an S3 bucket and/or already get image optimization through something like https://imgproxy.net.

**Reply by emilio-martinez** [reply]
Agreed. I'm aligned in principal to identifying common issues and problems, but using img hardly seems inherently problematic. The error would need to occur in a very specific set of scenarios where it would clearly be beneficial — otherwise it's just noise.

**Reply by housseindjirdeh** [reply]
This is great feedback, thank you both 🙏 

I think the main argument here is that I think in the majority of cases, using `<Image>` would result in a net performance benefit than placing  an `<img>` tag. There are always going to be scenarios where folks would be creating their own custom `<Image>` component and/or handling image optimization themselves for which they can disable the rule, but I expect this to be the minority.

CC @atcastle in case he has some thoughts on this.

**Comment by correttojs** [comment]
what about a rule to check the functions exported from within a page? I.e. only allow `default`, `getStaticProps`, `getStaticPaths` ,...
These cannot be type checked so a linting rule would be nice

**Reply by housseindjirdeh** [reply]
That sounds like a great place to have a rule in place for (to display as a warning at least but potentially even as an error?). My only concern is that I wonder if this changes too frequently as new APIs get included and others become deprecated. 🤔

CC @rgabs

**Comment by GorvGoyl** [comment]
`eslint-config-next`

> This repository has been archived by the owner. It is now read-only.
Last update: 4 years ago.

why are we considering an abandoned config :(

**Reply by karlhorky** [reply]
Not abandoned: https://www.npmjs.com/package/eslint-config-next/v/10.2.1-canary.3

The package is here on GitHub: https://github.com/vercel/next.js/tree/canary/packages/eslint-config-next

**Reply by GorvGoyl** [reply]
ah, author mentioned the old url for github on npm  (https://github.com/mvila/eslint-config-next)
![image](https://user-images.githubusercontent.com/7106086/117800922-50b4b200-b271-11eb-97f8-3f0d87011db7.png)

**Reply by karlhorky** [reply]
Interesting, this is something that shouldn't be happening... in [the actual published `package.json` file](https://unpkg.com/browse/eslint-config-next@10.2.1-canary.3/package.json), there is no mention of this `mvila`....

<img width="1253" alt="Screen Shot 2021-05-11 at 12 39 33" src="https://user-images.githubusercontent.com/1935696/117802776-2a354d80-b256-11eb-9a41-7968fcf38c3e.png">

<img width="982" alt="Screen Shot 2021-05-11 at 12 39 41" src="https://user-images.githubusercontent.com/1935696/117802785-2d303e00-b256-11eb-836f-a7d1faf0b0b9.png">

cc @Timer @timneutkens

**Reply by timneutkens** [reply]
We took over the package and it's currently on **canary** so it's entirely expected that npm is not displaying the contents of the package yet. https://www.npmjs.com/package/eslint-config-next/v/10.2.1-canary.5 is published correctly and has the latest code. Once we do a **stable** release the package will be showing the new contents.

**Reply by karlhorky** [reply]
Ah, I wasn't aware that it's possible that the current contents of canary releases may not be displayed on npm's website, interesting! 🤔

**Reply by timneutkens** [reply]
There's a lot to say about the npm website 😅

**Reply by housseindjirdeh** [reply]
In case it helps, I wrote full steps to try the feature out in https://github.com/vercel/next.js/discussions/25008 :)

**Comment by neefrehman** [comment]
It would be great if `eslint-config-next` could also bundle `eslint-plugin-jsx-a11y` and extend it's recommended ruleset, like with the other plugins.

**Reply by mryechkin** [reply]
@neefrehman agreed - though one thing to watch out for with `eslint-plugin-jsx-a11y` is that it will throw errors with the default usage of `next/link` component. You would need to override a couple rules, as suggested in [this issue](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/402#issuecomment-368305051):

```jsx
  'jsx-a11y/href-no-hash': 0,
  'jsx-a11y/anchor-is-valid': [
    'error',
    {
      components: ['Link'],
      specialLink: ['hrefLeft', 'hrefRight'],
      aspects: ['invalidHref', 'preferButton'],
    },
  ]
```

This is what I'm doing in my own [custom ESLint config](https://www.npmjs.com/package/eslint-config-acme) currently, and it's working well so far

**Reply by neefrehman** [reply]
Ah yes good point! forgot about this. All the better reason to add this rule to the config then :)

**Reply by tubbo** [reply]
This was the biggest bummer when we added `jsx-a11y` to our ESLint config, mostly having to turn this rule off. I would love to see some built-in parser support that can not only detect that a `<Link><a>...</a></Link>` component is the right move, but also that stuff like `passHref` is being used in cases where it needs to be, etc. One could really improve the experience and get some a11y-compliancy to boot.

**Reply by housseindjirdeh** [reply]
@JeffersonBledsoe recently opened #25770 to add some initial rules from `jsx-a11y`, but as warnings. Although I would love to bundle the recommended ruleset directly within `eslint-config-next`, I worry that it may be too invasive to begin with in regards to failing `next build` (since mostly all rules are defined as errors) and would result in folks in removing the config entirely.

@neefrehman Would you be okay with us including it in a phased approach as warnings before considering the recommended ruleset?

**Reply by housseindjirdeh** [reply]
With regards to overriding certain rules to play nicely with `next/link`, is `anchor-is-valid` the only rule that needs to be modified? I would be happy to include the workaround listed above by @mryechkin until Next.js resolves the need to include `<a>` tags with `<Link>` ([RFC](https://github.com/vercel/next.js/discussions/8207)), but I also see other options are recommended in the [docs](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/36102cd271194ddc9dc9ec0c0e78ceb13482ff68/docs/rules/anchor-is-valid.md#case-i-use-nextjs-and-im-getting-this-error-inside-of-links).

@tubbo We now have a [rule](https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/lib/rules/link-passhref.js) that warns you if `passHref` is not used when it should be. If there are any other specifics we can check for, please feel free to open an issue :)

**Reply by neefrehman** [reply]
>Would you be okay with us including it in a phased approach as warnings before considering the recommended ruleset?

Yes I'm happy with a phased approach here.

>is `anchor-is-valid` the only rule that needs to be modified?

In my case that is the only one I've had to modify

**Comment by GorvGoyl** [comment]
can we inform an unaware user if they install plugins seperately which were already installed using config-next? afaik underlying plugins from config-next are not shown in project package.json or eslint file so it's unclear for a user to know which plugins does it extend.

**Reply by housseindjirdeh** [reply]
ESLint, and consequently `next lint` and `next build`, will warn you if there are conflicting plugins:

```
Plugin "react" was conflicted between ".eslintrc » plugin:react/recommended" and ".eslintrc » eslint-config-next".
```

I've also added a [section](https://nextjs.org/docs/basic-features/eslint#migrating-existing-config) to the docs that recommends adding the plugin directly instead of `eslint-config-next` if you already have a configuration defined. I'll be updating this to be more specific. Hopefully this all suffices, or would you still like Next.js to show a warning if you install an already used plugin? :)

**Reply by SalahAdDin** [reply]
It is not clear for me, if we use `"plugin:@next/next/recommended"` at `extends` as the documentations says, does it include `next/core-web-vitals`?

Thanks

**Reply by leochiu-a** [reply]
@SalahAdDin No, `plugin:@next/next/recommended` will not include `next/core-web-vitals`. If you really need `next/core-web-vitals`, you should only use `next/core-web-vitals` at `extends` instead. `next/core-web-vitals` already extends `plugin:@next/next/recommended`.

**Reply by SalahAdDin** [reply]
So, for getting also core web vitals `plugin:@next/next/core-web-vitals` is fair enough, right?
